### PR TITLE
SignIn with Discoverable Credentials

### PR DESCRIPTION
**Pull request by xifengfang-okta**
## Description:
- Add `Sign in with security key or biometrics` button
- Add security key or biometric enrollment link when webauthn failed

Screen Shot 2022-08-18 at 10 52 25 AM

Screen Shot 2022-08-18 at 10 52 12 AM

Eng-OnePager:
https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2668265730/Biometric+for+CIAM+Eng+Design+-+Phase+1
https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2682880307/Biometric+for+CIAM+Eng+Design+-+Phase+1+-+sign-in+with+enrollment

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

- Did you verify the change by running downstream monolith artifacts? (YES | NO | UNSURE)

### Reviewers:
@okta/ciamx 

### Issue:

- [OKTA-511555](https://oktainc.atlassian.net/browse/OKTA-511555)
- [OKTA-518908](https://oktainc.atlassian.net/browse/OKTA-518908)

### CODE REVIEW

File: test/unit/spec/v2/view-builder/views/ChallengeWebauthnView_spec.js [hunk 1]
```diff
@@ -297,4 +300,102 @@ describe('v2/view-builder/views/webauthn/ChallengeWebauthnView', function() {
...
+        signature: 'magizh',
+        userHandle: 'userId',
+      },
+    };
+    jest.spyOn(navigator.credentials, 'get').mockReturnValue(Promise.resolve(assertion));
+    testContext.init(
+      ChallengeWebauthnResponse.currentAuthenticator.value,
+      [],
+      {},
+      {canEnrollResidentKey:true});
```
Comment (justinabrokwah-okta) [inline test/unit/spec/v2/view-builder/views/ChallengeWebauthnView_spec.js:377]: nit: spacing comment again - Bacon may complain about it

File: test/unit/spec/v2/view-builder/views/ChallengeWebauthnView_spec.js [hunk 2]
```diff
@@ -297,4 +300,102 @@ describe('v2/view-builder/views/webauthn/ChallengeWebauthnView', function() {
...
+    jest.spyOn(BrowserFeatures, 'isSafari').mockReturnValue(false);
+    jest.spyOn(navigator.credentials, 'get').mockReturnValue(Promise.reject({
+      message: 'error from browser',
+      name: 'NotAllowedError',
+    }));
+    testContext.init(
+      ChallengeWebauthnResponse.currentAuthenticator.value,
+      [],
+      {},
+      {canEnrollResidentKey:true});
```
Comment (justinabrokwah-okta) [inline test/unit/spec/v2/view-builder/views/ChallengeWebauthnView_spec.js:393]: nit: spacing

File: test/unit/spec/v2/view-builder/views/IdentifierView_spec.js [hunk 3]
```diff
@@ -295,4 +298,18 @@ describe('v2/view-builder/views/IdentifierView', function() {
     expect($usernameError.text()).toEqual('Username is required!');
     expect($psswordError.text()).toEqual('Password is required!');
   });
+
+  it('should show signin with biometric button when launch-webauthn-authenticator remediation exist', function() {
+    jest.spyOn(AppState.prototype, 'hasRemediationObject').mockImplementation(remediation => {
+      if(remediation === FORMS.LAUNCH_WEBAUTHN_AUTHENTICATOR) {
+        return true;
+      }
+      return false;
```
Comment (justinabrokwah-okta) [inline test/unit/spec/v2/view-builder/views/IdentifierView_spec.js:307]: nit: one-liner
`return remediation === FORMS.LAUNCH_WEBAUTHN_AUTHENTICATOR`

**Review by justinabrokwah-okta — COMMENTED** [review]
PR looks pretty good overall, left a few nit comments

File: src/v2/view-builder/views/webauthn/ChallengeWebauthnView.js [hunk 4]
```diff
@@ -88,12 +93,20 @@ const Body = BaseForm.extend({
...
     }).then((assertion) => {
+      const credentials = {
+        clientData: CryptoUtil.binToStr(assertion.response.clientDataJSON),
+        authenticatorData: CryptoUtil.binToStr(assertion.response.authenticatorData),
+        signatureData: CryptoUtil.binToStr(assertion.response.signature)
+      };
+      const hasUserHandleSchema = this.options.appState.getSchemaByName('credentials.userHandle');
+      if (hasUserHandleSchema) {
+        _.extend(credentials, {
+          userHandle: CryptoUtil.binToStr(assertion.response.userHandle ?? '')
```
Comment (vicentevillegas-okta) [inline src/v2/view-builder/views/webauthn/ChallengeWebauthnView.js:104]: out of curiosity. what does it mean this line?  specially the double 
`??` `CryptoUtil.binToStr(assertion.response.userHandle ?? '')`.

Comment (xifengfang-okta) [reply src/v2/view-builder/views/webauthn/ChallengeWebauthnView.js:104]: according to webauthn spec, userHandle is nullable, so I use this `nullish coalescing operator` to handle null case

Comment (vicentevillegas-okta) [reply src/v2/view-builder/views/webauthn/ChallengeWebauthnView.js:104]: awesome, I just learned a new operator in JS, thank you @xifengfang-okta

File: src/v2/view-builder/views/webauthn/EnrollWebauthnView.js [hunk 5]
```diff
@@ -60,14 +60,18 @@ const Body = BaseForm.extend({
     const relatesToObject = this.options.currentViewState.relatesTo;
     const activationData = relatesToObject?.value.contextualData.activationData;
     if (webauthn.isNewApiAvailable()) {
+      const excludeCredentials = activationData.authenticatorSelection?.requireResidentKey === true ?
+        [] :
+        getExcludeCredentials(this.options.appState.get('authenticatorEnrollments').value);
```
Comment (justinabrokwah-okta) [inline src/v2/view-builder/views/webauthn/EnrollWebauthnView.js:65]: Can we do an optional chain on value? 
`?.value`

Comment (xifengfang-okta) [reply src/v2/view-builder/views/webauthn/EnrollWebauthnView.js:65]: this enrollment flow is after use authenticated, it appears that `this.options.appState.get('authenticatorEnrollments')` will not be null

File: assets/sass/v2/_webauthn.scss [hunk 6]
```diff
@@ -21,5 +21,13 @@
     .idx-webauthn-enroll-text {
       margin-bottom: 20px;
     }
+
+    .setup-webauthn-residentkey-text {
+      margin-bottom: 25px;
+      .setup-webauthn-residentkey-link {
+        color: $primary-button-bg-color;
+        padding-left: 5px;
```
Comment (alexdahl-okta) [inline assets/sass/v2/_webauthn.scss:29]: 👋 Any reason this link is getting a special (background?) color and padding?

Exceptions can make it more difficult for customers to theme and customize the widget.

Comment (xifengfang-okta) [reply assets/sass/v2/_webauthn.scss]: good catch, padding is not necessary, color is needed but changed to the common `link-text-color`, thanks a lot

**Review by aarongranick-okta — APPROVED** [review]

File: packages/@okta/i18n/src/properties/login.properties [hunk 7]
```diff
@@ -601,6 +601,7 @@ primaryauth.newUser.tooltip.close=Close
 oktaVerify.description = To access this resource, your organization requires you to sign in with Okta FastPass.
 oktaVerify.appDescription = To access {0}, your organization requires you to sign in with Okta FastPass.
 oktaVerify.button = Sign in with Okta FastPass
+signinWithWebAuthn.button = Sign in with security key or biometric
```
Comment (justinabrokwah-okta) [inline packages/@okta/i18n/src/properties/login.properties:604]: Just confirming, did you get the new properties looked at?

**Review by justinabrokwah-okta — APPROVED** [review]
LGTM 🚀 , just left a question about new strings in `login.properties`

File: packages/@okta/i18n/src/properties/login.properties (hunk 7 above)
Comment (xifengfang-okta) [reply packages/@okta/i18n/src/properties/login.properties:604]: @IanHakes-okta  can you please take a look at these two text?

**Review by IanHakes-okta — CHANGES_REQUESTED** [review]
Two fixes to the login properties (which also need to updated in the leakthrough properties files).

File: packages/@okta/i18n/src/properties/login.properties (hunk 7 above)
Comment (xifengfang-okta) [reply packages/@okta/i18n/src/properties/login.properties]: updated

File: packages/@okta/i18n/src/properties/login.properties [hunk 8]
```diff
@@ -971,6 +972,7 @@ oie.verify.webauthn.cant.verify.enrollment.step1 = Open your Okta Dashboard (e.g
 oie.verify.webauthn.cant.verify.enrollment.step2 = Go to Settings > Security Methods
 oie.verify.webauthn.cant.verify.enrollment.step3 = On Okta Verify, click "Set up"
 oie.verify.webauthn.cant.verify.enrollment.step4 = Scan the QR code using Okta Verify and follow instructions to finish enrolling your account
+oie.enroll.webauthn.rk.link = If you don't have Security key or Biometric set up, <a href="#" class="setup-webauthn-residentkey-link">finish setting up by signing in.</a>
```
[the remark this answers is not in the document]
Comment (xifengfang-okta) [reply packages/@okta/i18n/src/properties/login.properties]: updated

**Review by IanHakes-okta — APPROVED** [review]
Strings LGTM

File: packages/@okta/i18n/src/properties/login_ok_PL.properties [hunk 9]
```diff
@@ -712,6 +713,7 @@ oie.verify.webauthn.cant.verify.enrollment.step1 = 》Öþéñ ýöûŕ Öķţå
 oie.verify.webauthn.cant.verify.enrollment.step2 = 》Ĝö ţö Šéţţîñĝš ≥ Šéçûŕîţý Ṁéţĥöðš Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.verify.webauthn.cant.verify.enrollment.step3 = 》Öñ Öķţå Ṽéŕîƒý، çļîçķ ″Šéţ ûþ″ 䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.verify.webauthn.cant.verify.enrollment.step4 = 》Šçåñ ţĥé ǪŔ çöðé ûšîñĝ Öķţå Ṽéŕîƒý åñð ƒöļļöŵ îñšţŕûçţîöñš ţö ƒîñîšĥ éñŕöļļîñĝ ýöûŕ åççöûñţ 䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
+oie.enroll.webauthn.rk.link = 》Îƒ ýöû ĥåṽéñ´ţ šéţ ûþ ýöûŕ šéçûŕîţý ķéý öŕ ƀîöɱéţŕîç åûţĥéñţîçåţîöñ، ýöû çåñ çöɱþļéţé ţĥé þŕöçéšš åƒţéŕ ýöû šîĝñ îñ ûšîñĝ ≤å ĥŕéƒ≂″ €홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ#″ çļåšš≂″šéţûþ‐ŵéƀåûţĥñ‐ŕéšîðéñţķéý‐ļîñķ″≥ţĥîš ļîñķ≤⁄å≥· โ⾼ئ䀕ヸ€홝한Ӝฐโ《
```
Comment (zihanwang-okta) [inline packages/@okta/i18n/src/properties/login_ok_PL.properties:716]: Seems the html tag is not properly translated. 
It is a different story, but we might need to upgrade the pseudo-loc @okta/i18n.pseudo-loc package version.

**Review by zihanwang-okta — APPROVED** [review]
The changes in properties files lgtm.

**Pull request merged**